### ISSUE TITLE
JacksonDeserializer with claimTypeMap throws NullPointerException while parsing

### ISSUE THREAD

**Issue by Nirostar** [jwtk/jjwt#564]
I have a problem with parsing my JWS Token with jjwt. The following code snippet

```Java
JacksonDeserializer<Map<String, ?>> deserializer = new JacksonDeserializer<>(Map.of("user", JwtUser.class));
JwtParser parser = Jwts.parserBuilder()
        .setSigningKey(key)
        .deserializeJsonWith(deserializer)
        .build();
Jws<Claims> jws = parser.parseClaimsJws(token);
```
throws the following NullPointerException.
```Stacktrace
java.lang.NullPointerException: null
	at java.base/java.util.ImmutableCollections$Map1.containsKey(ImmutableCollections.java:742) ~[na:na]
	at java.base/java.util.Collections$UnmodifiableMap.containsKey(Collections.java:1450) ~[na:na]
	at io.jsonwebtoken.jackson.io.JacksonDeserializer$MappedTypeDeserializer.deserialize(JacksonDeserializer.java:123) ~[jjwt-jackson-0.11.0.jar:0.11.0]
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4202) ~[jackson-databind-2.10.2.jar:2.10.2]
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3266) ~[jackson-databind-2.10.2.jar:2.10.2]
	at io.jsonwebtoken.jackson.io.JacksonDeserializer.readValue(JacksonDeserializer.java:104) ~[jjwt-jackson-0.11.0.jar:0.11.0]
	at io.jsonwebtoken.jackson.io.JacksonDeserializer.deserialize(JacksonDeserializer.java:96) ~[jjwt-jackson-0.11.0.jar:0.11.0]
	at io.jsonwebtoken.impl.DefaultJwtParser.readValue(DefaultJwtParser.java:613) ~[jjwt-impl-0.11.0.jar:0.11.0]
	at io.jsonwebtoken.impl.DefaultJwtParser.parse(DefaultJwtParser.java:308) ~[jjwt-impl-0.11.0.jar:0.11.0]
	at io.jsonwebtoken.impl.DefaultJwtParser.parse(DefaultJwtParser.java:541) ~[jjwt-impl-0.11.0.jar:0.11.0]
	at io.jsonwebtoken.impl.DefaultJwtParser.parseClaimsJws(DefaultJwtParser.java:601) ~[jjwt-impl-0.11.0.jar:0.11.0]
	at io.jsonwebtoken.impl.ImmutableJwtParser.parseClaimsJws(ImmutableJwtParser.java:173) ~[jjwt-impl-0.11.0.jar:0.11.0]
```

I debugged it and saw, that the NPE happened while parsing the header of the Token which is looking like this:

```JSON
{"alg":"HS256"}
```

The NPE happens, because the `parser.currentName()` in `JacksonDeserializer::deserialize(JsonParser parser, DeserializationContext context)` is `null` and therefore it can't look it up in the claimTypeMap. The currentName is null because it parses the root element and doesn't have a current name.

Parsing my Object by querying the map values and parsing it works, but is very ugly. I would rather use the above mentioned method.

```Java
JacksonDeserializer<Map<String, ?>> deserializer = new JacksonDeserializer<>();
JwtParser parser = Jwts.parserBuilder()
        .setSigningKey(key)
        .deserializeJsonWith(deserializer)
        .build();
Jws<Claims> jws = parser.parseClaimsJws(token);
Claims claims = jws.getBody();
Map<String, Object> user = (Map<String, Object>) claims.get("user");
JwtUser jwtUser = new JwtUser();
jwtUser.setPersonId((Long) user.get("personId"));
jwtUser.setUsername((String) user.get("username"));
jwtUser.setAuthorities((Collection<String>) user.get("authorities"));
return jwtUser;
```

Maybe it suffices to check whether the parser.currentName() is null and don't look it up if it is.

**Comment by lhazlewood** [comment]
@Nirostar thanks very much for reporting this!  We'll get a test case for this and release a fix in 0.11.1.

cc @bdemers

**Comment by bdemers** [comment]
@Nirostar Thanks for providing great info, it made tracking this down much easier (though I still banged my head on this for a bit).

It turns out the Java 9+`Map.of(...)` implementation creates a map that does NOT allow for checking `containsKey(null)`, other map implementations DO allow for this.
The short term workaround is to use the `io.jsonwebtoken.lang.Maps.of` the utility class provided with by JJWT (or another map impl) for example:

```java
Map.of("user", JwtUser).build()
```

We will get the null check added so this doesn't trip anyone else up! Thanks again for the detailed report!

**Comment by lhazlewood** [comment]
@bdemers did https://github.com/jwtk/jjwt/pull/566 resolve this issue?

**Comment by bdemers** [comment]
Yup! fixed in: #566
